libs/header/p2p/exchange.go

```diff
@@ -264,7 +265,7 @@ func (ex *Exchange[H]) request(
 	log.Debugw("requesting peer", "peer", to)
 	responses, size, duration, err := sendMessage(ctx, ex.host, to, ex.protocolID, req)
 	ex.metrics.observeResponse(ctx, size, duration, err)
-	if err != nil {
+	if err != nil && err != io.EOF {
```

Comment:
Maybe really swap out io.EOF with nil to avoid changing such ifs all over the place?

Reply:
and in metrics as well

Reply:
> Maybe really swap out io.EOF with nil to avoid changing such ifs all over the place?

You mean here? Yesterday, I thought that it's not correct that sendMessage decides what error is allowed... So I decided to return any kind of error from the sendMessage and the caller will decide what and how he will handle it

Reply:
> that sendMessage decides what error is allowed.

I haven't said anything like this, or why do you see sendMessage should not decide this?

Reply:
There can be a case where EOF is not a valid error, e.g. Responsibilities of sendMessage are:
* send request;
* return response/error;

And caller should decide how to handle responses/errors, imo

Reply:
Then Maybe it should not do the EOF check for logs there?

Reply:
for these particular cases, we can mark eof as a valid response.

Reply:
regarding the check, I can reset it after sendMessage, but it will look weird. So, I think the best place to keep is really inside sendMessage.

Reply:
👌🏻